File: src-java/techniques/VendingMachine.java
package techniques;

import acm.program.ConsoleProgram;

import java.util.*;

public class VendingMachine extends ConsoleProgram {

    public void initChangeMoney(TreeMap<Integer, Integer> coins) {
        coins.put(5, 3);  // three 5ct coins
        coins.put(10, 2);  // two 10ct coins
        coins.put(20, 3);  // ..
        coins.put(50, 2);
        coins.put(100, 2);
        coins.put(200, 1);
    }

    public void giveChangeGreedy(TreeMap<Integer, Integer> coins, int amount) {
        for (int coin: coins.keySet()) {
            int nrOfCoins = coins.get(coin);
            while((nrOfCoins > 0) && ((amount - coin) >= 0)) {
                amount = amount - coin;
                nrOfCoins--;
                print(coin + ", ");
            }
            coins.put(coin, nrOfCoins);
        }
    }

    public void run() {
        TreeMap<Integer, Integer> coins = new TreeMap<>(Collections.reverseOrder());

        initChangeMoney(coins);

        // just counting the number of coins, we do not care what their value is
        int numCoins = 0;
        for (int val: coins.values())
            numCoins += val;

        println("Coins: " + coins + "\n");
        println("Number of coins: " + numCoins);
        println("Brute force:");
        double numCombinations = Math.pow(2, numCoins);
        println("Number of combinations: " + numCombinations);
        println("Takes roughly (1ns / calculation): " + (numCombinations / (1000 * 1000)) + " sec");

        // so brute force is no option, let's have a look at greedy
        giveChangeGreedy(coins, 85);
    }
}

File: src-python/9-Techniques/VendingMachine.py
# We've learned that binary search trees are highly efficient data structures for
# maintaining ordered data. For our vending machine implementation, it would be
# very helpful if our dictionary maintained its keys in descending order. Java
# provides a built-in data structure called TreeMap for this purpose.
# Does Python offer something similar?

# Let's ask an LLM: "I need a size-ordered dictionary in Python. What data structure
# is efficient? Give an example."

# In Python, the built-in dict does not maintain the keys in sorted order — it maintains
# insertion order (since Python 3.7+). If you need a dictionary with keys automatically
# kept in sorted order, the most efficient solution is:
#   - use SortedDict from the sortedcontainers library
#   - keys stay automatically sorted.
#   - efficient operations: O(log n) for insert, delete, and lookup.
# Install it: pip install sortedcontainers
from sortedcontainers import SortedDict

class VendingMachine:
    def init_change_money(self, coins: list[int]) -> SortedDict:
        coins_desc = SortedDict(lambda x: -x)

        for coin in coins:
            coins_desc[coin] = coins_desc[coin] + 1 if coin in coins_desc else 1

        print(coins_desc)
        return coins_desc

    def give_change_greedy(self, coins: SortedDict, amount: int):
        for coin in coins.keys():
            nr_of_coins = coins[coin]
            while nr_of_coins > 0 and (amount - coin) >= 0:
                amount -= coin
                nr_of_coins -= 1
                print(coin, end=", ")
            coins[coin] = nr_of_coins

    def run(self):
        coins = [5, 5, 5, 10, 10, 20, 20, 20, 50, 50, 100, 100, 200]
        coins_sorted = self.init_change_money(coins)

        # just counting the number of coins, we do not care what their value is
        num_coins = sum(coins_sorted.values())

        print("Coins:", coins, "\n")
        print("Number of coins:", num_coins)
        print("Brute force:")
        num_combinations = 2 ** num_coins
        print("Number of combinations:", num_combinations)
        print("Takes roughly (1ns / calculation):", num_combinations / (1000 * 1000), "sec")

        # so brute force is no option, let's have a look at greedy
        print("Greedy:")
        self.give_change_greedy(coins_sorted, 85)

if __name__ == "__main__":
    vm = VendingMachine()
    vm.run()